# Sreality
### 12.3.2020
https://www.sreality.cz/hledani/

* Další odkazy:

https://www.realingo.cz/

https://pronajem.zvladneme.cz/

https://www.bezrealitky.cz/



In [ ]:
from IPython.display import HTML

HTML('''<script>
code_show=true; 
function code_toggle() {
 if (code_show){
 $('div.input').hide();
 } else {
 $('div.input').show();
 }
 code_show = !code_show
} 
$( document ).ready(code_toggle);
</script>
<form action="javascript:code_toggle()"><input type="submit" value="Ukaž/Schovej všechen kód"></form>''')

-----------------------
## Progress:
##### 12.3.2020 - start


-----------------------
## Progress:
##### 12.3.2020 - start
* Nefunguje beautiful soup,
* nefunguje  XML
* ani dlouhý i krátký formáty div / span atd.
* nefunguje ani google na h2 tagy, tři návody  k ničemu

##### 18.3. - D. mě zachránil se Seleniem, zapomněl jsem na scrapping dynamické JS stránky
##### 19.3. - Scrapnuto všech 600 stran, parametr na počet stran, na byt/dům, prodej/pronájem, změněno H2 na celý popisek a scrapp again
##### 20.3. - Cleaning - split jména, 6+ pokojů na metry
##### 21.3. - Cleaning - separace města z lokality, Praha extra, nějaký testovací groupbies, pivot table
###### - poučka 1 - "list of index out of range" nejspíš znamená že indexy neexistují
###### - poučka 2 - skoro vše se dá rozkouskovat a printnout
##### 22.3. - zkoumání okresů a krajů, pokus o scrapp okresů a obcí
##### 23.3. - Wiki - nalezeny kraje, okresy a obce (města i vsi) - už jen scrap
##### 24.3. - Státní správa - vymyšlen scrapp a úpravy podle české abecedy!
##### 25.3. - Vzdávám scrapp obcí, mám EXCEL od Z. Manuálně vecpáno všechno do Prahy. Stažena fresh data, reakce na airbnb fail? Jinak mám vylepšené gorupbys, Median, První ukázka klukům
##### 27.3. - Čistá data Obce-Okresy_kraje, hrál jsem si s hledáním stringů ve strinzích, geopy od Lukáše + JSON otestováno funkční hledání souřadnic
##### 28.3. - motám se trochu v souřadnicích a pomalém stahování
##### 29.3. - pomalé stahování pokořeno, mám 12.000 souřadnic, JSON upraven na stahování i ceny, upravení detekce adres z loakce souřadnice,odhad stahování adres = 10h
##### 30.3 - pomalý download - mám 6.000/12.000 přes den. Objeveny tony a první znělka z Beepů. Error handling - nekonečný repeat funkce = stažení zbytku přes noc :)
###### - poučka 3 - dlouhé kody kouskuj a ošetři proti erroru, vždycky komentuj proces PRINTy, přidej zvukové efekty, průěbžně ukládej
###### - poučka 4 - schovat všechna data. Pak už jen nové či staré inzeráty spáruju podle názvu (adresa, místnosti, metry) nebo přes coords, cokoliv krom coords_to_adress
##### 31.3. - stáhnuty všechny adresy, upravena datam chybí už jen METRY. Zopakovány základní analýzy na nových datech
##### 1.4. - Staženy všechny plochy za 50 min (po odblokování jejich zpomalovače). Upravena struktura práce. Odstraněny taré sekce analýzy. Udělán kus čištění, např. plochy
###### - poučka 5 - na jeden sloupec = apply(), když má vzniknout víc sloupců tak map() ??
###### - poučka 6 - STRIP(x) odřeže ze stringu znaky X, SPLIT(X) rozseká strink na místech X
##### 2.4. - uděláno čištění měst - celkem 4 problémy, hezky uhlazené, rozhodnutí co dropovat. Už se to blíží....
##### 3.4. - Dokončeno čištění dat, přidání obyvatel, kontroly a Analýza naostro
##### 4.4. - studuji si publikování notebooku, vychytávky, cool barploty, uděláno první PREZI a posláno D. a L. - feedback je SVG obrázky a Cena za metr !
###### poučka 7 - za nehezké sloupečky jde přidat .to_frame()
###### poučka 8 - můžu dělat conditional formatting: .style.applymap(fce)
##### 5.4. - s D. pořešen hezký export do HTML, dále interaktivní skrývání kodových buněk
##### 6.4. PAUZA
##### 7.4. - Jenom duplikáty, rozmýšlení co dropnout
##### 8.4. - opět řeším duplikáty..
##### 9.4. - PAUZA
##### 10.4. - Další scrapping, po 12 dnech, budu moci zkontrolovat ID jestli se na nich staly změny, kolik přibylo, koik ubylo
##### 11.4. - PAUZA
##### 12.4. - Automatizace scrapingu
##### 13.4. - Stažení nových dat + Skoro hotové mapování Coords na již získané adresy
###### poučka 9 - concat je blbé na sloupce, musím mapovat spíš přes merge, nechat jeden sloupec shodně pojmenovaný a ostatní co chci přidat, pak nejsou duplikátní sloupce
##### -
##### 16.4. boj se stahováním nových adres, potíže s připojením, zkoumání parametrů GEOPY
##### 17.4. po mnoha testech nakoenc zůstanu u "kabelu v obýváku", rychlost 8-20x lepší než kabel v ložnici:D
##### 18.4. Udělán další SCrap dat, zásek u JSON - cena a další, obejevena chyba u Pronájem DOmu - stažena jen jedna stránka. Jinak dodělán merge starýcha nových adres a implementace do funkce
##### 19.4. Dodělán Merge Adres, Nové schéma v malování na hezkou návaznost kroků, udělána funkce na update adresáře, dostažen JSON na předvčerejší data, nalezen problém s duplikovanými názvy měst v Adresáři ve dvou formátech
##### 20.4. Adresář očištěn o duplicity Okresů a Krajů, všude odmazány mezery před jmény, upravena funkce Update_Adressy a Map_old_Adres aby ukládala jen unikátní kombiance url_id + short_coords, totéž při loadingu - konečně snad sjednocené, přehledné, bez duplikování. Dokončena Fáze 1 = scrapping včetně testu
##### 21.4. CELANING - udělány adresy, namapován počet obyvatel, updatování databáze včetně správného počtu řádků, žádné špatné mazání duplikátů, Dále Plocha přes regex - zvládá i "1 000", Ceny /1000 a Cena za metr. HOTOVA fáze 2 !!

## To do:

### NUTNÉ:
##### ZAUTOMATIZOVAT PROCES NA SCRAP + úpravy PRODEJ BYTY KOMPLET, ODENDAT MANUÁLKY, VYMYSLET DOMÝŠLENÍ ADRES ZE SOUŘADNIC
##### Proovnat CENY ZA METR jako pprůměr oblasti, HEATMAPY?, TABLEAU/POWERBI, SVG OBRÁZKY DO PREZI (?)

* zjisti jestli staré i nové inzeráty co jen změní cenu mají stejné ID .- dost by to pomohlo
* ## zaokrouhlování short_coords kvůli hrubšímu odhadování města ??
* najít duplikáty - viz 3 byty co mají téměř nejdražší cenu za metr jsou prakticky stejné - podle plochy + pokojů + města + oblasti ??
* nebo anopak podle stejných bytů, co se jen liší v ceně hledat změny cen bytů, pokud by to teda nešlo přes to ID
* K duplikátům - právě jsem našel ?problém? se psoustou bytů an stejné souřadnici - konkrétně Olomouc. vadí to něčemu? hůř se pak hledají duplikáty - což nemusí vadit ale uvidíme
* Nebo Lukášův návrh dělat něco jakop prmůěry cen za oblast (za metr) a na základě toho sledovat vývoje cen. Problém je taky v tom, že inzeráty taky mizí takže nepůjdou porovnat 
* x
* x
* Dodělat zbylé 1-3 varianty celaningu zbylých 3 kombinací - prodej, domy ... problém s adresami je jiný
* Mergnout asi klidně dohromady
* pracovat s metry a velikostmi, nějaké histogramy, shluky

* ---------------
### ESTETICKÉ
* naklikat ovládací prvky na filtry a gráfy a hlavně ty gruppy
* Highlight pivot extremy https://pandas.pydata.org/pandas-docs/stable/user_guide/style.html
* geodata, mapa s kraji?
* jak vyřešit změny v čase? jednak přímo během stahování, jednak dlouhodobě?
* pomocí URL se ještě rozhodnout vyřadit ze seznamu Dražby prodejů
* vyřešit záhadné byty v Troji (podle Realingo), kdy jeden byt se skládá ze 4 bytů skoro stejných

* ---------------
### PROPAGACE
* nájemní byty - sepsat situaci (ceny, počtyobyvatel, počty v nájmech atd - viz starý Forbes),
* pak pozitiva, pak negativa, pak happy scénář, pak možnosti úvěrů, pak ytb: 63 story

-----------------------

# Obsah

1. Importování Packagů
2. Scraping
3. Rozšíření datasetu
4. Čištění
5. Analýzy

-----------------------
# 1. Importování Packagů
-----------------------

In [ ]:

##### Obecné ############
import pandas as pd # for dataframe
from pandas import DataFrame
import numpy as np # for arrays
import matplotlib as plt
from matplotlib.pyplot import figure     # for saving and changing size of plots

from collections import Counter             # Na počítání prvků v dimenzi  
from datetime import datetime
import re                    # relativní Novinka - regular expressions
from time import sleep # for sleeping (slowing down) inside a function
import random  # for random number sleeping
import math    # Zaokrouhlování
import time    # Měření scrappingu délka spánků

import itertools   # Tool na odlistování vnořených listů



##### David a Scraping ############
import requests # for robots check
from bs4 import BeautifulSoup
from selenium import webdriver  

##### Lukáš a GeoPy ############        # Odsud dolů samé novinky:
from geopy.geocoders import Nominatim   # Geolocator   # pip install geopy  # Co je geopandas???
import json
from geopy.exc import GeocoderTimedOut

##### Widgets ############
import ipywidgets as widgets
from ipywidgets import interact, interact_manual
from IPython.display import display
import os
from IPython.display import Image

##### Bonus - Hudba ############
import winsound

##### Vizualizace ############
import seaborn as sns



-----------------------
# 2. Scraping
-----------------------

In [ ]:
print(requests.get("https://www.sreality.cz/robots.txt").text)

## 2a) Scraping 1 - Původně: Soup se vším krom Souřadnic
* Nedostatek - nemám souřadnice - > nemám pořádné adresy -> nemám dobrá města, okresy, kraje

In [ ]:

# scrappig elements, které v sobě mají Lokaci, Velikost, Cenu

def get_soup_elements(typ_obchodu = "prodej", typ_stavby = "byty", pages = 5):  
    
    browser = webdriver.Chrome()
    
    ##########################################
    # 1. Volba Prodej/Pronájem, Byty/Domy,                 --Aukce/Bez Aukce (jen pro Prodeje) zatím nechávám být, cpe se mi to doprostřed url
    ##########################################    
    
    url_x = r"https://www.sreality.cz/hledani"             # Všechny byty, lepší než naklikat všechno, kratší url a přehlednější
    url = url_x + "/" +  typ_obchodu + "/" +  typ_stavby
    
    ##########################################
    # 2. načtení webu
    ##########################################
    browser.get(url) # (url).text ??
    sleep(random.uniform(1.0, 1.5))
    innerHTML = browser.execute_script("return document.body.innerHTML")
    soup = BeautifulSoup(innerHTML,'lxml') # "parser" ??
    elements = soup.find_all(class_ = "text-wrap")     # !!!!!!!!!!!!!!!!! změněno, protože H2 neobsahovalo všechny věci, jen nadpisek.
    
    ##########################################
    # 3. zjištění počtu listů - mělo by být optional, ale nevadí
    ##########################################
    records = soup.find_all(class_ ='numero ng-binding')[1].text
    records = re.split(r'\D', str(records))                         # Tady byla chyba, kdoví proč "pages" místo "records", což mi to podělávalo
    records = ",".join(records).replace(",", "")
    records = int(records)
    max_page = math.ceil(records / 20)   # 12 072 inzerátů  = 604 stránek, správně
    print("----------------")
    print("Scrapping: " + str(typ_obchodu) + " " + str(typ_stavby))
    print("Total records: " + str(records))
    print("Total pages: " + str(max_page))
    
    ##########################################
    # 4. nastavení počtu stránek  -mělo být víc promakané
    ##########################################
    if pages == 999:      # (NE)Speciální případ, chci všechny inzeráty - standardně tu bude asi kolem 600 stránek max, takže 999 je volné k použití
        pages = max_page
    
    print("I scrap (only) " + str(pages) + " pages.")
    print("----------------")
    
    ##########################################
    # 4. Scrapping zbylých listů - naštěstí v jednom okně
    ##########################################    
    
    for i in range(pages-1):   # pages = 650 -> range-1 = (0-648) -> i+2 = (2-650) ->  pages = 2-650
        i = i+2
        
        print("Page " + str(i-1) + " = " + str(round(100*(i-1)/(pages), 2)) + "% progress.")
        print("Estimated time remaining: " + str(round(random.uniform(3.3, 3.85)*(pages-(i-1)), 2 )) + " seconds.")  # čas kolem 3.5, spíš maličko nahoru. Odhadnuto podle Scrappingu 600 stránek
        
        url2 = url + "?strana=" + str(i)
        browser.get(url2)

        sleep(random.uniform(1.0, 1.5))

        innerHTML = browser.execute_script("return document.body.innerHTML")
        soup2 = BeautifulSoup(innerHTML,'lxml') # "parser" ??
        elements2 = soup2.find_all(class_ = "text-wrap")    # !!!!!!!!!!!!!!!!! změněno, protože H2 neobsahovalo všechny věci, jen nadpisek.
        
        elements = elements + elements2
        # tyto se už můžou posčítat, naštěstí, řpedtím než z nich budeme dělat elems = prvky třeba jména
        
    browser.quit()   # novinka :)
    
    return elements

######################################################
#######################################################

"""
x = 999
elements_prodej_byty = get_soup_elements(typ_obchodu = "prodej", typ_stavby = "byty", pages = x)
elements_pronajem_byty = get_soup_elements(typ_obchodu = "pronajem", typ_stavby = "byty", pages = x)
elements_prodej_domy = get_soup_elements(typ_obchodu = "prodej", typ_stavby = "domy", pages = x)
elements_pronajem_domy = get_soup_elements(typ_obchodu = "pronajem", typ_stavby = "domy", pages = x)
"""
#

## 2b) Scraping 2 - href odkazy => JSON
* souřadnice jsou zdlouhavé, ale lepší info
* JSON je příjemný na pickování informací

In [ ]:

# Nový scraping, místo elements, které v sobě mají Lokaci, Velikost, Cenu
# stahuju Href link, kde je číslo inzerátu, přes ten půjdu na JSON a na souřadnice mimojiné


def get_soup_elements(typ_obchodu = "prodej", typ_stavby = "byty", pages = 1):  
    
    browser = webdriver.Chrome()
    
    url_x = r"https://www.sreality.cz/hledani"             # Všechny byty, lepší než naklikat všechno, kratší url a přehlednější
    url = url_x + "/" +  typ_obchodu + "/" +  typ_stavby

    browser.get(url) # (url).text ??
    sleep(random.uniform(1.0, 1.5))
    innerHTML = browser.execute_script("return document.body.innerHTML")
    soup = BeautifulSoup(innerHTML,'lxml') # "parser" ??
    
    ##### ELEMENTS = Všechny linky s číslem inzerátu, jsou nejprve zdvojené
    elements = []    
    
    for link in soup.findAll('a', attrs={'href': re.compile("^/detail/")}):     ##############################################
        link = link.get('href')   ##############################################
        elements.append(link)     ##############################################
    elements = elements[0::2]     ##############################################


    ##########################################
    # 3. zjištění počtu listů - mělo by být optional, ale nevadí
    ##########################################
    records = soup.find_all(class_ ='numero ng-binding')[1].text
    records = re.split(r'\D', str(records))                         # Tady byla chyba, kdoví proč "pages" místo "records", což mi to podělávalo
    records = ",".join(records).replace(",", "")
    records = int(records)
    max_page = math.ceil(records / 20)   # 12 072 inzerátů  = 604 stránek, správně
    print("----------------")
    print("Scrapping: " + str(typ_obchodu) + " " + str(typ_stavby))
    print("Total records: " + str(records))
    print("Total pages: " + str(max_page))
    
    ##########################################
    # 4. nastavení počtu stránek  -mělo být víc promakané
    ##########################################
    if pages == 999:      # (NE)Speciální případ, chci všechny inzeráty - standardně tu bude asi kolem 600 stránek max, takže 999 je volné k použití
        pages = max_page
    
    print("I scrap (only) " + str(pages) + " pages.")
    print("----------------")
    
    ##########################################
    # 4. Scrapping zbylých listů - naštěstí v jednom okně
    ##########################################    
    
    for i in range(pages-1):   # pages = 650 -> range-1 = (0-648) -> i+2 = (2-650) ->  pages = 2-650
        i = i+2
        
        sys.stdout.write('\r'+ "Page " + str(i-1) + " = " + str(round(100*(i-1)/(pages), 2)) + "% progress. Remaining: " + str(round(random.uniform(3.3, 3.85)*(pages-(i-1)), 2 )) + " seconds.") 

        url2 = url + "?strana=" + str(i)
        browser.get(url2)

        sleep(random.uniform(1.0, 1.5))

        innerHTML = browser.execute_script("return document.body.innerHTML")
        soup2 = BeautifulSoup(innerHTML,'lxml') # "parser" ??
        
        elements2 = []
        
        for link in soup2.findAll('a', attrs={'href': re.compile("^/detail/prodej/")}):  ##############################################
            link = link.get('href') ##############################################
            elements2.append(link)  ##############################################
   
        elements2 = elements2[0::2]  ##############################################
        
        elements = elements + elements2

    
    browser.quit()   # novinka :)
    
    return elements

######################################################
#######################################################

"""
x = 999
elements_prodej_byty = get_soup_elements(typ_obchodu = "prodej", typ_stavby = "byty", pages = x)
elements_pronajem_byty = get_soup_elements(typ_obchodu = "pronajem", typ_stavby = "byty", pages = x)
elements_prodej_domy = get_soup_elements(typ_obchodu = "prodej", typ_stavby = "domy", pages = x)
elements_pronajem_domy = get_soup_elements(typ_obchodu = "pronajem", typ_stavby = "domy", pages = x) 
"""
#

## 2c) Získání URL_IDs - dál už jen prodej byty
* elements uložím jako data

In [131]:
elements = pd.DataFrame({"url":elements_prodej_byty})

def get_id(x):
    x = x.split("/")[-1]
    return x

elements["url_id"] = elements["url"].apply(get_id)
elements.head()
#"""
elements.to_excel(r"URL_data_prodej_byty.xlsx")
#"""
#

## 2d) Získání Coordinates + Cena + Plocha z JSON
* COORDS mám z JSON
* Cena a Plocha přidána taky z JSON
* "Jméno" a z něj rozměry a byt/prodej dům/pronájem můžu z URL co už mám

In [ ]:
def get_coords_price_meters(x):
    
    url = "https://www.sreality.cz/api/cs/v2/estates/" + str(x)
    
    response = requests.get(url)
    byt = json.loads(response.text, encoding = 'UTF-8')                                
    try:
        coords = (byt["map"]["lat"], byt["map"]["lon"])
    except:
        coords = (0.01, 0.01)
    try:
        price = byt["price_czk"]["value_raw"] 
    except:
        price = -1
        
    try:
        description = byt["meta_description"]
    except:
        description = -1
    
    return coords, price, description

##################################################xx
"""
start = time.time()

elements["coords"], elements["cena"], elements["popis"] = zip(*elements['url_id'].map(get_coords_price_meters))
# TOTO JE NA Tripple STAHOVÁNÍ

#data["popis"] = data["url_id"].apply(get_coords_price_meters)   
# TOTO JE SINGLE STAHOVÁNÍ

end = time.time()
duration = end-start
display(duration)
display(time.time())
#
"""
# Naposledy 10.4. aplikováno na 13.000 záznamů: 66 minut

In [46]:
elements

url      url_id
0      /detail/prodej/byt/2+1/zlin-zlin-na-honech-ii/...   861945436
1      /detail/prodej/byt/4+1/zlin-zlin-dr--kolarika/...   726089308
2      /detail/prodej/byt/1+kk/praha-kosire-hlavackov...  1831546460
3      /detail/prodej/byt/3+kk/praha-smichov-holeckov...  1074892380
4      /detail/prodej/byt/1+kk/praha-smichov-holeckov...   528649820
5      /detail/prodej/byt/3+kk/praha-smichov-graficka...  2923327836
6      /detail/prodej/byt/1+kk/praha-smichov-graficka...  3268153692
7      /detail/prodej/byt/5+kk/praha-smichov-holeckov...  4047618396
8      /detail/prodej/byt/2+kk/praha-kamyk-rabynska/2...  2421223004
9      /detail/prodej/byt/4+kk/ostrava-hermanice-herm...  1192685148
10     /detail/prodej/byt/2+kk/praha-nusle-hradeckych...  2368417372
11     /detail/prodej/byt/1+kk/praha-letnany-kadeckov...  1228664412
12     /detail/prodej/byt/3+1/libcice-nad-vltavou-lib...  2272219484
13     /detail/prodej/byt/2+1/praha-nusle-na-vysledku...  1915432540
14     /detail/prodej/byt/2+kk/praha-zizkov-ke-kapslo...  1646997084
15     /detail/prodej/byt/2+1/chodov-chodov-cs--odboj...  2720738908
16     /detail/prodej/byt/4+1/praha-zizkov-baranova/3...   375471708
17     /detail/prodej/byt/3+kk/praha-zizkov-baranova/...  1365327452
18     /detail/prodej/byt/3+kk/praha-zabehlice-chocer...  1297170012
19     /detail/prodej/byt/3+kk/praha-modrany-vorarska...  1745108572
20     /detail/prodej/byt/2+1/praha-zabehlice-bojanov...  3252764252
21     /detail/prodej/byt/3+kk/chyne-chyne-ecksteinov...  2469080668
22     /detail/prodej/byt/2+1/hodonin-hodonin-/327438...  3274387036
23           /detail/prodej/byt/3+kk/klicany--/539700828   539700828
24     /detail/prodej/byt/4+kk/praha-modrany-vorarska...  3149057628
25     /detail/prodej/byt/3+kk/slavkov-u-brna-slavkov...  3048578652
26     /detail/prodej/byt/3+kk/ivancice-ivancice-bozi...  1316015708
27     /detail/prodej/byt/3+1/karlovy-vary-karlovy-va...  1613442652
28     /detail/prodej/byt/1+1/chomutov-chomutov-kamen...  1345007196
29     /detail/prodej/byt/2+kk/usti-nad-labem-klise-m...  1268072028
...                                                  ...         ...
13258  /detail/prodej/byt/3+1/karlovy-vary-karlovy-va...  3609444700
13259  /detail/prodej/byt/2+kk/karlovy-vary-karlovy-v...  1379778908
13260  /detail/prodej/byt/3+1/karlovy-vary-karlovy-va...   652132700
13261  /detail/prodej/byt/1+kk/vaclavov-u-bruntalu-ho...  2975993948
13262  /detail/prodej/byt/2+kk/holasovice-holasovice-...  1358479708
13263  /detail/prodej/byt/2+kk/holasovice-holasovice-...  3103310172
13264  /detail/prodej/byt/2+kk/karlovy-vary-dvory-che...  2445664604
13265  /detail/prodej/byt/3+kk/karlovy-vary-dvory-che...  2245386588
13266  /detail/prodej/byt/2+kk/karlovy-vary-karlovy-v...  2116149596
13267  /detail/prodej/byt/1+1/karlovy-vary-karlovy-va...   678617436
13268  /detail/prodej/byt/2+1/karlovy-vary-karlovy-va...  2140397916
13269  /detail/prodej/byt/3+1/karlovy-vary-karlovy-va...  2727600476
13270  /detail/prodej/byt/4+kk/praha-nove-mesto-tyrso...    41804124
13271  /detail/prodej/byt/2+kk/karlovy-vary-karlovy-v...   336519516
13272  /detail/prodej/byt/2+1/karlovy-vary-karlovy-va...  1416552796
13273  /detail/prodej/byt/3+kk/karlovy-vary-karlovy-v...   801759580
13274  /detail/prodej/byt/5+1/praha-stare-mesto-karlo...   780398940
13275  /detail/prodej/byt/2+1/karlovy-vary-karlovy-va...  2715910492
13276  /detail/prodej/byt/3+1/karlovy-vary-karlovy-va...  2823655772
13277  /detail/prodej/byt/2+1/karlovy-vary-cast-obce-...  2271125852
13278  /detail/prodej/byt/4+1/marianske-lazne-marians...   378384732
13279  /detail/prodej/byt/2+kk/karlovy-vary-karlovy-v...  1831711068
13280  /detail/prodej/byt/3+1/karlovy-vary-karlovy-va...  2517610844
13281  /detail/prodej/byt/2+kk/marianske-lazne-marian...  2382369116
13282  /detail/prodej/byt/3+1/karlovy-vary-cast-obce-...   688808028
13283  /detail/prodej/byt/4+1/karlovy-vary-karlovy-va...  1896677468
13284  /detail/prodej/byt/2+kk/ostrava-moravska-ost

### Problémy při downloadu - každý druhý pokus tohle:

HTTPSConnectionPool(host='www.sreality.cz', port=443): Max retries exceeded with url:
        /api/cs/v2/estates/4054269532 (Caused by NewConnectionError('<urllib3.connection.VerifiedHTTPSConnection object at 0x0000020502133978>:
        Failed to establish a new connection: [WinError 10051] A socket operation was attempted to an unreachable network',))

-----------------------
# 3. Rozšíření datasetu
-----------------------

## 3a) Pro Scrapping 1 - Popis, Lokalita, cena

In [ ]:

def data_from_elements(elements):    # Jediné řešní, jak každému setu vyrobit jeho proměnné, je funkce a assignovat ji datasetům
    
    name = []   # typ, dům, místnosti, metry ...
    locality = []  # město + oblast
    price = []  # cena
    
    for elem in elements:
        element1 = elem.find(class_ ='name ng-binding')   # class_ místo class  !!!!!
        name.append(element1.text)
        
        element2 = elem.find(class_ ='locality ng-binding')   
        locality.append(element2.text)
        
        element3 = elem.find(class_ ='norm-price ng-binding') 
        price.append(element3.text)

    display(len(name))
    display(len(locality))    
    display(len(price))    
    
    data = pd.DataFrame({'popis': name, 'lokalita': locality, 'cena': price})
    
    def Amount_arrange(x):
        x = re.findall(r'\d+', str(x))
        x = ''.join(x)
        # x.fillna(0)
        return x

    data["cena"] = data["cena"].map(Amount_arrange)
    
    return data
    
######################################################
#######################################################
"""
data_prodej_byty = data_from_elements(elements_prodej_byty)
data_pronajem_byty = data_from_elements(elements_pronajem_byty)
data_prodej_domy = data_from_elements(elements_prodej_domy)
data_pronajem_domy = data_from_elements(elements_pronajem_domy)
#"""
#-----------------------
# SAVED
#-----------------------
"""
data_prodej_byty.to_excel("data_prodej_byty.xlsx")
data_pronajem_byty.to_excel("data_pronajem_byty.xlsx")
data_prodej_domy.to_excel("data_prodej_domy.xlsx")
data_pronajem_domy.to_excel("data_pronajem_domy.xlsx")
"""
# .....

# 3b) Coordinates -> Adresa  // Získání z předchozízch
#### = 10h čistého stahování // Získání

In [ ]:

def adresa(x):

    geolocator = Nominatim(timeout = 3, user_agent = "specify_your_app_name_here")
    location = geolocator.reverse(x.strip("())"))  #database_Kladno_rent["Souřadnice"][0].strip("())"))
    #display(location)    

    ### Pojistky, kdyby náhodou chybělo nějaké pole - asi by stačila jediná podmínka u Ulice....
    try:
        ulice  = location[0].split(",")[-7]
    except:
        ulice  = -1
    try:
        město = location[0].split(",")[-6]
    except:
        město  = -1
    try:
        okres = location[0].split(",")[-5]
    except:
        okres  = -1
    try:
        kraj = location[0].split(",")[-4]  
    except:
        kraj  = -1       

    return ulice, město, okres, kraj
##################################################################

### REPETICE - pomohlo překonat časté odpojování od serverů - dokud se neobjevila druhá chyba, ale i tak

def repeat_adress():
    try:
        data["oblast"], data["město"],  data["okres"] ,  data["kraj"]  = zip(*data['coords'].map(adresa))
    except GeocoderTimedOut:
        print("Another try")
        data["oblast"], data["město"],  data["okres"] ,  data["kraj"]  = zip(*data['coords'].map(adresa))
        
##################################################################

"""
data["oblast"], data["město"],  data["okres"] ,  data["kraj"]  = zip(*data['coords'].map(adresa))

####

repeat_adress()

"""
#

### Hudební vsuvka za odměnu

In [ ]:
import winsound
import time 

sec = 300  # milliseconds - for Beep
half_sec = 150
pause = 1   # Seconds - for Sleep

# záhadně to nebere násobky floatem ani dělení integerem i když ot ve výsledku je celé číslo

# Hz
A3 = 220
C4 = 262   #261.63
D4 = 294   #293.67
E4 = 330   #329.63
F4 = 350   #349.23
G4 = 392
A4 = 440
C5 = 523   #523.25

koef = 2

def we_are_the_champions():
     
    winsound.Beep(koef*F4, 4*sec)
    winsound.Beep(koef*E4, sec)
    winsound.Beep(koef*F4, sec)
    winsound.Beep(koef*E4, 3*sec)
    winsound.Beep(koef*C4, 2*sec)
    winsound.Beep(koef*A3, 3*half_sec)
    winsound.Beep(koef*D4, 3*half_sec)
    winsound.Beep(koef*A3, 10*sec)
    time.sleep(pause)
    winsound.Beep(koef*C4, sec)
    winsound.Beep(koef*F4, 4*sec)
    winsound.Beep(koef*G4, 2*sec)
    winsound.Beep(koef*A4, 2*sec)
    winsound.Beep(koef*C5, 4*sec)
    winsound.Beep(koef*A4, 2*sec)
    winsound.Beep(koef*D4, sec)
    winsound.Beep(koef*E4, sec)
    winsound.Beep(koef*D4, 6*sec)

    
    
we_are_the_champions()

## 3c) Charakteristiky získané z URL
* Byt/ dům
* Prodej/ Pronájem
* Plocha, Plocha na metr

In [ ]:
# Prodej / pronájem
# Byt / dům
# Počet pokojů
# => To vše jde z URL splitem

def characteristics(x):
    x = x.split("/")
    buy_rent = x[2]
    home_house = x[3]
    rooms = x[4]
    # meters = ....
    
    return buy_rent, home_house, rooms

##################################################################

display (data["url"].iloc[2]) #.map(name)

data["prodej"], data["dům"],  data["pokoje"] = zip(*data['url'].map(characteristics))
data.head()



# Plocha
# Vysosnuto z popisku

def plocha(x):
    try:
        metry = x.split()[2]
    except:
        metry = -1
    return metry

##################################################################

data["plocha"] = data['popis'].map(plocha)


# Cena za metr !! - jde až zpětně, až vyčistím Plochu !

data["cena_metr"] = round(data['cena']/data["plocha"], 1)

#

## 3d) Obce, Okresy, Kraje, obyvatelstvo

### 1) Státní správa
    https://www.statnisprava.cz/rstsp/ciselniky.nsf/i/CZ0201
    - problémy s abecedou, Brno-město nemá obce atd
### 2) Wiki
    https://cs.wikipedia.org/wiki/Seznam_katastr%C3%A1ln%C3%ADch_%C3%BAzem%C3%AD_v_okrese_Bene%C5%A1ov
    - taky celý dost na houby, nezačal jsem
### 3) Volby.cz
    https://www.volby.cz/pls/kv2018/kv31?xjazyk=CZ&xid=1
    - Bylo by hezké, jsou to tables, easy to scrapp, jen jiný počet než v 1) (75 vs 77 okresů)
### 4) Staťák - excel
    https://www.czso.cz/csu/czso/pocet-obyvatel-v-obcich-za0wri436p?fbclid=IwAR1haSspuynZB8Awn08WDriMkUcsUCz4fHH9Pw2CwMDVGHPGJERxaqbrVg8
    - Tohle asi bude top, Zuzka mě zachránila


* 14 krajů
* 77 okresů podle státní správy včetně PRAHA
* 6258 obcí a újezdů  K 27. květnu 2016 (ČSÚ)  


* Chrudim, Cheb, Chomutov … co to je a kde to je? :D				
* Totéž Pelhřimov, pardubice, Prachatice…				
* Přerov a Prostějov…


In [ ]:
# Použito pro pozdější namapování chybějících krajů v sekci 4d)

obce = pd.read_excel("obce_okresy_kraje.xlsx")

def first_lower(x):
    if len(x) == 0:
        return x
    else:
        return x[0].lower() + x[1:]

# Kvůli sjednocení s "data"
obce["Okres"] = obce["Okres"].apply(first_lower)

############# 
# Přidání ovytelstva: viz 4d)
obce.head()

-----------------------
# 4. Čištění
-----------------------

* Toto vše je verze pro prodeje bytů, 
* Další bude pro pronájem domů - jiná struktura

## 4a) Pro Scrapping 1 - čištění dat

In [ ]:

def cleaning(data):


    ##########################################
    # 1. Split názvu na jednotlivé parametry
    ##########################################
    data[['buy_rent','home_house', "rooms", "meters", "m","0", "1", "2", "3"]] =  data.name.str.split(expand=True)  # Vytvoří se 4 idiotské sloupce navíc, nevím proč
    data = data[["name", "locality", "price", 'buy_rent','home_house', "rooms", "meters"]]

    ##########################################
    # 2. Dropnutí nulových cen
    ##########################################
    display(len(data))
    data["price"] = data["price"].fillna(0)    #
    data = data.loc[~(data.price ==0)]
    display(len(data)) 

    
    data["price"] = (data["price"]/1000).astype(int)    
    # Převedení millionů na tisíce - snad přehlednější
    
    ##########################################
    # 3. 6+ Pokojů doplnění metrů z názvu (odhaleno díky Unique hodnotám - 40x "pokojů")
    ##########################################
    def six(x):
        x = re.findall(r'\d{3}', str(x))
        x = ''.join(x)
        return x

    data.loc[data["rooms"] == "6", "meters"] = data.loc[data["rooms"] == "6", "name"].map(six)
    # Komplikované, ale praktické
    
    ##########################################
    # 4. Města, Okresy
    ##########################################
  
    # Potřebuji detekovat čárku s mezerou a pokračovat buĎ k mezeře s pomlčkou nebo ke konci stringu... (víceslovná města)
    def city(x):
        x = re.split(r'\W\s', x)   # (W + s) == non-letter(-) + space // případně by to chtělo (s + W) == space + (-) a definovatten znak, aby to nebralo tečky !!!!!!
        try:
            x = x[1]    # TOTO jsem chtěl ,ale některé adresy mají jen jednu část - Město (snad)
        except:
            x = x[0]
        return x
    
    data["city"] = data["locality"].map(city)
    
    # Pomocný sloupec kde beru jen první slovo z názvu "city", což mi zajistí slovo Praha místo Praha_X
    def praha(x):
        x = re.findall(r'\w+', x)[0]
        return x
    
    data["city_one_praha"] = data["city"].map(praha)
    # A tady dokopíruju ostatní názvy jiné než Praha, a mám hotovyý sloupec
    data.loc[data["city_one_praha"] != "Praha", "city_one_praha"] = data["city"]

    
    
    return data

######################################################
#######################################################
data = cleaning(data)
data
# 12.440 -> 11.599

## 4b) Úprava cen na tisíce

In [19]:
data["cena"] = data["cena"].apply(lambda x: int(round(x/1000, 0)))

# Zmenší mi dataset z 1532 na 1558 KB :)

## 4c) Plocha - částečně MANUÁLNÍ úpravy !!

In [ ]:
# Problémy s plochou - některé inzeráty se očividně změnily,
# už nejde přs ůvodní URL dohledat JSON info, 
# např index 195 - Studénka, už existuje pod jiným ID
# Celkem 500 bez plochy - musím je ale nutně vyhazovat? nerad bych, když vím aspoň dispozici

data.groupby('plocha')['url'].nunique()
# ZÁHADA:
# Před uložením excelu se tohle řadilo -1, 0, 100, 1000, 11, 110, ...
# Nyní po novém načtení z excelu to jde hezky -1, 13, ... 1900

data[data["plocha"] == 1]

In [57]:
### PLOCHY 

data.groupby('plocha')['popis'].nunique()

# Jeden záznam - UPRAV SLOUPEC PLOCHA!!   !!!!!!!!!!!!!!!!!!!!!!!
"""
data.plocha.iloc[2523] = 1
data.plocha.iloc[3249] = 1105
data.plocha.iloc[10440] = 1000
"""

#######################################################

# 123 záznamů Plocha == DISPOZICE
# Snaha:
# dataframe.SLOUPEC[řádky z df které:[sloupec] == "něco"].iloc[číslo řádku].rozděl()[druhý prvek]
#data.popis[data["plocha"] == "dispozice"].iloc[0].split()[3]   - # iloc se získá APPLY

def dispozice_metry(x):
    x = x.split()[3]
    return x

# Měním sloupec PLOCHA          # Používám sloupec POPIS                              # Používám fci na 3 řádky
data.plocha[data["plocha"] == "dispozice"] = data.popis[data["plocha"] == "dispozice"].apply(dispozice_metry)

# KONTROLA:
data[data["pokoje"] == "atypicky"]

# Jeden nový záznam - UPRAV SLOUPEC PLOCHA!!!!!!!!!!!!!!!!!!!!!!!!!!!
"""
data.plocha[data["plocha"] == "1"] = 1900
"""

#######################################################

# 34 záznamů Plocha == POKOJŮ

def dispozice_metry2(x):
    x = x.split()[5]
    return x

# Měním sloupec PLOCHA          # Používám sloupec POPIS                              # Používám fci na 3 řádky
data.plocha[data["plocha"] == "pokojů"] = data.popis[data["plocha"] == "pokojů"].apply(dispozice_metry2)

#KONTROLA:
#data[data["pokoje"] == "6-a-vice"]

C:\Users\jiri.zelenka\AppData\Local\Continuum\anaconda3\lib\site-packages\pandas\core\indexing.py:189: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self._setitem_with_indexer(indexer, value)
C:\Users\jiri.zelenka\AppData\Local\Continuum\anaconda3\lib\site-packages\ipykernel_launcher.py:24: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
C:\Users\jiri.zelenka\AppData\Local\Continuum\anaconda3\lib\site-packages\ipykernel_launcher.py:43: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy


## 4d) Města + Obyvatelstvo - částečně MANUÁLNÍ úpravy !!

In [ ]:

data.groupby('město')['popis'].nunique()

###### Problém 1  - mám u krajů a možná okresů před jménem mezeru
def uprav_adresy(x):
    x = re.findall(r'\w+', str(x))
    x = ' '.join(x)
    return x
data["kraj"] = data["kraj"].apply(uprav_adresy)
data["okres"] = data["okres"].apply(uprav_adresy)
data["město"] = data["město"].apply(uprav_adresy)
data["oblast"] = data["oblast"].apply(uprav_adresy)
# Pozn: akorát mi to změnilo Oblasti, města, okresy i Kraje z -1 na 1


###### Problém 2 - některé adresy jsou beznadějně prázdné - smažu 55 kusů
display(len(data[data["coords"] == "(0.01, 0.01)"]))
data[data["coords"] == "(0.01, 0.01)"]
# HOTOVO !! smazáno v cleaningu



In [ ]:

###### Problém 3 - 59 řádků má posunuté Město - okres - kraj !!!
data[data["oblast"] == "1" ].head(10)

def kraj_check(x):
    y = x.split()[0]
    return y

data["kraj_check"] = data["kraj"].apply(kraj_check)


# U těch řádků, kde byl okres v názvu kraje posuň všechny tři hodnoty "doleva"
data.oblast[data["kraj_check"] == "okres"] = data.město[data["kraj_check"] == "okres"]
data.město[data["kraj_check"] == "okres"] = data.okres[data["kraj_check"] == "okres"]
data.okres[data["kraj_check"] == "okres"] = data.kraj[data["kraj_check"] == "okres"]
data.kraj[data["kraj_check"] == "okres"] = -1

# Už jen u Krajů doplnit správný kraj podle okresů (59 kusů)
display(len(data[data["kraj_check"] == "okres"]))
data[data["kraj_check"] == "okres"]

# Oříznutý datase OBCE
obce_kus = obce[["Okres", "Kraj"]]
obce_kus.columns = ["okres", "kraj"]
obce_kus.head()


# LEFT Merge by měl přiřadit okresům správné Kraje z df obcí
data = pd.merge(data, obce_kus, on=['okres'], how = "left")
display(len(data))

# Tohle by mělo pořešit problém s duplikáty - ale sebralo mi to nakonec dva řádky navíc, které jsou ofiko duplikáty...
data.drop_duplicates(inplace = True)

display(len(data))

# Asi zbytečnost, ale funguje to - přesun části sloupce
def nový_kraj(x):
    x = x
    return x

data.kraj_x[data["kraj_check"] == "okres"] = data.kraj_y[data["kraj_check"] == "okres"].apply(nový_kraj)

# Přejmenování a získání lepších sloupců
data = data.rename(columns={'kraj_x': 'kraj'})
data = data[["popis", "prodej", "dům", "pokoje", "plocha", "oblast", "město", "okres", "kraj", "cena", "url", "url_id", "coords"]]

data[data["kraj_check"] == "okres"]


In [ ]:
###### Problém 4 - pár divných měst s čísly zmizí při cleaningu - OK
#                  - pár oblastí jsou jen čísla - to ničemu nevadí, dá se říci. Zatím nechávám.
#                  - 19 vadných okresů se smaže při celaningu - OK
#                   - 19 krajů totéž - OK
#                   - Vysočina  kraj spojit s Kraj Vysočina asi roznito při domapování z Excelu

data.kraj[data["kraj"] == "Vysočina kraj"] = "Kraj Vysočina"


# Závěrečné kontroly:
 ## Lepší hledat Unique přes URL než POPIS, ten je častí prázdný = 1 a zkresluje počet záznamů

"""
data.groupby('oblast')['url'].nunique()
data.groupby('město')['url'].nunique()
data.groupby('okres')['url'].nunique()    # 77 okresů
data.groupby('kraj')['url'].nunique()      # 14 krajů
"""
#


In [ ]:
####### Problém 5:
#       Přidávání obyvatelstva je konečně možné (viz sekce 3d))

město_obyv = obce[["Název_obce", "Okres", "Počet_celkem"]]
město_obyv.columns = ["město", "okres", "Počet_celkem"]

# Musím použít města i okresy, abych se vyhnul duplikacím - obce se stejnými názvy (Např. Tachov) 
data = pd.merge(data, město_obyv, on=["město", "okres"], how = "left")
display(len(data))

# Počet po odmazání duplikátů odpovídá
data.drop_duplicates(inplace = True)
len(data)


## 4e) Dropnutí Bez Cen // ((Plochy)) // Pokojů // Města // Kraje // Okresu

* Asi zvlášť problémy s geografií, a plochou pak celaring !!!

*  ze 12.498:
* CENA : 897 je bez ceny (k dostání u RK) - Smažu
* PLOCHA: 503 je bez plochy - nemazal bych, není nutná
* POKOJE - ty jsou všude :)
* ADRESY  55 nemá ani kraj, okres, nic (chybí coordinates, i když by šlo z popisku) - Smažu
* Bez ceny a be z Coordinates se celkem překrývá, celkem 908

* CELKEM 12,498 -> ?? kompletních záznamů

In [ ]:
print("Celkem záznamů bez ceny: "+ str( len(data[data["cena"] == 0]) )   )
print("Celkem záznamů bez plochy: "+ str( len(data[data["plocha"] == -1]) )  +". Nemažu je." )
print("Celkem záznamů bez pokojů: "+ str( len(data[data["pokoje"] == -1]) )   )
print("Celkem záznamů bez oblasti a okresu: "+ str(len(data[data["coords"] == "(0.01, 0.01)"])))

x = len(data)

def clearing(data):

    data = data[data["cena"] != 0]   # Pozn - bylo -1, ale pak se dělilo tisícem a zaokrouhlovalo.
    display(len(data))
    #data = data[data["plocha"] != -1]
    #display(len(data))
    data = data[data["coords"] != "(0.01, 0.01)"]
    display(len(data))
   
    
    
    print("Finální délka datasetu: " + str(len(data)) + ", smazáno " + str(x-len(data)) + " záznamů.")
    
    return data
###############################################################

data = clearing(data)
data

## 4f) Dropnutí extrémů = pochybné záznamy, MANUÁLNĚ

In [ ]:
data = data[data["cena"] != 12]

x = data.index[data["cena"] == 320].tolist()  # Je jen jeden, naštěstí, jinak bych musel líp filtrovat
data  = data[data.index != x]

x = data.index[(data["cena"] == 180) & (data["město"] == "Praha")].tolist()  # Upřesněno na pražský byt
data  = data[data.index != x]
len(data)

## 4g) Pokus o hledání duplikátů


Shodnost ve všem kromě popisu a coords - 494 záznamů (2.239 bez shodnosti ceny)
        #3658	URL_ID: 1956396636	(50.071137083957, 14.4703357132463)	 - Míří na střed ulice
        #7096	URL_ID: 4064231004	(50.0724406380103, 14.4713951723744) - Míří na dům přímo	

Shodnost ve všem kromě Coords - 211 

Shodnost ve všem kromě popisu - 309
        #134   4096974428   - Popis Neratovice
        #140   1482702428   - Popis Předboj

Shodnost ve všem včetně včetně popisu - 189  
#####################################################################


Rozdíl Komplet identických včetně Ceny  bez ceny se liší o 19 záznamů, asi tedy smažu jen ty komplet stejné, včetně ceny, to by měla být tutovka !!

Druhý krok - koukám že popisy jsou někdy fakt jen drobnost, takže mažu i podle nich

Třetí krok - zbylo mi 20x2 záznamů, pozorování: Všechny dvojice mají téměř identické souřadnice - mažu taky !!

Otázka - dají se tři kroky nahradit jedním, tj smazat vše kde se neshoduje popisek ani cena (nebo jen jedno z toho)
to asi nevyjde stejně, protože to zabere i ty co se rozchází ve dvou věcech, a ty můžou být už opravdu jiné..?

In [ ]:

data_duplic = data[data.duplicated([  "prodej", "dům", "pokoje", "plocha","oblast", "město", "okres", "kraj", "cena", "coords"  ],keep= "first")]
#Mělo by být jedno, jestli definující sloupce dám dovnitř duplicated, nebo za druhé DATA, ale pak musím přidat[] navíc
# False -> Ponech si všechny zduplikované, "first" -> ponech si první výskyt

display(len(data_duplic))

display(data_duplic.sort_values("plocha").head(15))
# ZObrazí duplikované výsledky - pokud nahoře zvolím False, tak všechny výsledky se svými dvojníky

seznam = list(data_duplic.index)
data = data[ ~data.index.isin(seznam)]
display(len(data))

# !! Tímto ale smažu jen první výskyty z dat !

###############################################################################################################################################x
# Podruhé pro druhou skupinu shodností, nechci oboje spojit dohromady, protože bych povolil dvojitou chybu
# řekl bych že to je dobře. sice mi zbývá asi 150 potenciálních duplikátů, ale hned namátkově jsou ta mdva různé byty. Spíš to vypadá že se tam pricing řídí lokalitou a plochou, proto to má smysl chatne oba byty

data_duplic = data[data.duplicated([ "popis",  "prodej", "dům", "pokoje", "plocha","oblast", "město", "okres", "kraj", "cena"  ],keep= "first")]
display(len(data_duplic))

display(data_duplic.sort_values("plocha").head(15))

seznam = list(data_duplic.index)
data = data[ ~data.index.isin(seznam)]
display(len(data))

# 11.255 zbývá z 11.585 = 330 smazáno

## POznámka: data.iloc[1101] = 1101. záznam podle klasického řádkování
##           data.ix[1101] = prvek s idnexem 1101. Pokud ten nebyl resetován tak se to liší

## ??4h) Porovnání starých a nových inzerátů => DODĚLAT

In [ ]:
data_old = data_old[["cena", "pokoje", "plocha", "město"]]


mixed = pd.merge(data, data_old, on=["pokoje", "plocha", "město"], how = "inner")
display(len(mixed))
mixed

In [ ]:
mixed.drop_duplicates(inplace = True)
display(len(mixed))
mixed

In [169]:
data = pd.read_excel(r"data_prodej_byty_03_29.xlsx")
display(len(data))
data.head()

11585

popis  prodej  dům pokoje  \
0  Byt 4+kk 91 m² k prodeji U Stírky, Praha 8 - L...  prodej  byt   4+kk   
1  Byt 1+kk 34 m² k prodeji Pražská, Praha 10 - H...  prodej  byt   1+kk   
2  Byt 4+kk 91 m² k prodeji Praha 8; 7 190 000 Kč...  prodej  byt   4+kk   
3  Byt 1+kk 34 m² k prodeji Praha 10; 3 249 000 K...  prodej  byt   1+kk   
4  Byt 2+kk 72 m² k prodeji Blšanecká, Praha 10 -...  prodej  byt   2+kk   

   plocha     oblast  město                     okres                kraj  \
0      91      Libeň  Praha  okres Hlavní město Praha  Hlavní město Praha   
1      34   Hostivař  Praha  okres Hlavní město Praha  Hlavní město Praha   
2      91   Kobylisy  Praha  okres Hlavní město Praha  Hlavní město Praha   
3      34     Michle  Praha  okres Hlavní město Praha  Hlavní město Praha   
4      72  Uhříněves  Praha  okres Hlavní město Praha  Hlavní město Praha   

   cena  cena_metr                                                url  \
0  7190       79.0  /detail/prodej/byt/4+kk/praha-liben-u-stirky/2...   
1  3249       95.6  /detail/prodej/byt/1+kk/praha-hostivar-prazska...   
2  7190       79.0  /detail/prodej/byt/4+kk/praha-praha-8-/3489873500   
3  3249       95.6  /detail/prodej/byt/1+kk/praha-praha-10-/373560...   
4  5400       75.0  /detail/prodej/byt/2+kk/praha-uhrineves-blsane...   

       url_id                                coords  Počet_celkem  
0  2953002588  (50.1199607849121, 14.4618196487427)     1308632.0  
1   262721116  (50.0558700561523, 14.5230903625488)     1308632.0  
2  3489873500             (50.1243991667, 14.45457)     1308632.0  
3  3735604828                (50.063639, 14.473661)     1308632.0  
4  3234676316  (50.0330103411153, 14.5836153466492)     1308632.0

-----------------------
# 5. Analýzy
* https://pandas.pydata.org/pandas-docs/stable/user_guide/style.html
-----------------------


#### 0) Nadefinované grafické pomůcky
----
#### 1) Počty inzerátů
 * 1.a) Podle místností
 * 1.b) Podle měst/okresu/kraje
---
#### 2) Ceny
- 2.a) Podle místností
- 2.b) Podle měst
- - 2.b.1) Nejdražší města
- - 2.b.2) Nejlevnější města
- 2.c) Podle okresů
- - 2.c.1) Nejdražší okresy
- - 2.c.2) Nejlevnější okresy
- 2.d) Podle krajů
-----------------------
#### 3) Ceny podle místností a oblastí PRAHA & MOST
- 3.a) Ceny podle místností Praha
- 3.b) Ceny městských částí Prahy
- - 3.b.1) Nejdražší městské části
- - 3.b.2) Nejlevnější pražské části
- 3.c) Ceny podle místností Most
- 3.d) PRAHA vs MOST
-----------------------
#### 4) Pivot - Pražské části a pokoje
-----------------------
#### 5) Analýzy podle okresů a krajů = DODĚLAT
----
#### 6) Interaktivní widgety = DODĚLAT
---

### 0) Nadefinované grafické pomůcky

In [ ]:
### Barvičky do BARů

# červená a zelená jsou vidět na html exportu, modrá a rudá s černým písmem NE
# Takže na HTML export musím místo modré dát tyrkys...

zelená = "#22c02c"
tyrkys =   "#16c9ea"
modrá  =    "#0307ca"
červená =  "#e61717"
rudá =      "#a10000"
hnědá = "#4d0000"


# 1.) Zkusi barevnou škálu na celou tabulku. tím by se našly extrémy co nejdou podle tmavosti
# 2) Pivot - udělat jednak potlačení nul, a pak barevnou paletu?
# 3) Zkusit nějak aplikovat ten sloupec s dvěma arvami, ale musím mít kladné i záporné hodnoty, nějaké odchylky od průměru - to bych asi msuel třeba na všechny okresy najednou nebo něco


########################################################################

def small_red(val):
 
    color = 'red' if val < 500 else 'white'
    return 'color: %s' % color

 #.style.applymap(small_red)
########################################################################

def highlight_max(s):
 
    is_max = s == s.max()
    return ['background-color: green' if v else '' for v in is_max]

#  .style.apply(highlight_max) 

########################################################################
def highlight_median(s):   # Vyznačí nejvyšší hodnotu ve sloupci

    is_max = s == s.median()
    return ['background-color: green' if v else '' for v in is_max]

#  .style.apply(highlight_median) 


########################################################################
########################################################################

# https://jakevdp.github.io/PythonDataScienceHandbook/04.07-customizing-colorbars.html

"""
from matplotlib.colors import LinearSegmentedColormap

def grayscale_cmap(cmap):
    Return a grayscale version of the given colormap
    cmap = plt.cm.get_cmap(cmap)
    colors = cmap(np.arange(cmap.N))
    
    # convert RGBA to perceived grayscale luminance
    # cf. http://alienryderflex.com/hsp.html
    RGB_weight = [0.299, 0.587, 0.114]
    luminance = np.sqrt(np.dot(colors[:, :3] ** 2, RGB_weight))
    colors[:, :3] = luminance[:, np.newaxis]
        
    return LinearSegmentedColormap.from_list(cmap.name + "_gray", colors, cmap.N)
    

def view_colormap(cmap):
    Plot a colormap with its grayscale equivalent
    cmap = plt.cm.get_cmap(cmap)
    colors = cmap(np.arange(cmap.N))
    
    cmap = grayscale_cmap(cmap)
    grayscale = cmap(np.arange(cmap.N))
    
    fig, ax = plt.subplots(2, figsize=(6, 2),
                           subplot_kw=dict(xticks=[], yticks=[]))
    ax[0].imshow([colors], extent=[0, 10, 0, 1])
    ax[1].imshow([grayscale], extent=[0, 10, 0, 1])

view_colormap('viridis')
"""
#

In [ ]:
from IPython.display import HTML

HTML('''<script>
code_show=true; 
function code_toggle() {
 if (code_show){
 $('div.input').hide();
 } else {
 $('div.input').show();
 }
 code_show = !code_show
} 
$( document ).ready(code_toggle);
</script>
<form action="javascript:code_toggle()"><input type="submit" value="Ukaž/Schovej všechen kód"></form>''')

###  LOAD (or save)

In [ ]:
data = pd.read_excel(r"data_prodej_byty.xlsx")
print("Data mají " + str(len(data)) + " řádků." )

data_old = pd.read_excel(r"data_prodej_byty_OLD.xlsx")
print("Stará data mají " + str(len(data_old)) + " řádků. - Před pročištěním." )

obce = pd.read_excel("obce_okresy_kraje.xlsx")
print("Obce mají " + str(len(obce)) + " řádků." )


display(data.head())
display(data_old.head())
display(obce.head())

In [ ]:
data.info()

### 1) Počty inzerátů

### 1.a) Počty inzerátů podle místností

In [ ]:
# data.groupby('pokoje')['popis'].count().to_frame()  #.sum()

#data.pokoje.value_counts()
#data.pokoje.value_counts().to_frame()
data.pokoje.value_counts().to_frame().style.bar(align = "mid", color = tyrkys)


### 1.b) Počty inzerátů podle měst/okresu/kraje

In [ ]:
########################################################
# Třikrát identické funkce !! ale "nunique" hází jiný výsledek než "count" a "size" při kombinaci s "nlargest" !!!
########################################################
data.groupby('město')['popis'].nunique()
data.groupby('město')['popis'].count()   # COunt by mělo počítat všechny, nejen UNIQUE !!
data.groupby('město')['popis'].size()
########################################################

In [ ]:
# Všechna
data.groupby(['město', "okres"])['popis'].count().nlargest(10).to_frame().style.bar(align = "mid", color = modrá)

In [ ]:
# Bez Prahy
data[data["město"] != "Praha"].groupby(['město', "okres"])['popis'].count().nlargest(10).to_frame().style.bar(align = "mid", color = modrá)

# Nový kod jde připojit i za tohle :3
# Upraveno align na mid, aby ůstaly zachované délky sloupců

In [ ]:
# Všechny
data.groupby('okres')['popis'].count().nlargest(10).to_frame().style.bar(align = "mid", color = modrá)

In [ ]:
# Bez Prahy
data[data["město"] != "Praha"].groupby( "okres")['popis'].count().nlargest(10).to_frame().style.bar(align = "mid", color = modrá)

In [ ]:
data.groupby('kraj')['popis'].count().nlargest(20).to_frame().style.bar(align = "mid", color = modrá)

### 2) Ceny

### 2.a) Ceny podle místností

In [ ]:
data["cena"].agg(['min', "mean", np.median,'max', "count"]).apply(lambda x: round(x)).to_frame()


In [ ]:
data.groupby("pokoje")["cena"].agg(['min', "mean", np.median,'max', "count"]).apply(lambda x: round(x)).style.bar(subset = ["min", "mean", "median", "max"], align = "mid", color = rudá)
#data[data["plocha"] != -1].groupby("pokoje")["cena_metr"].agg(['min', "mean", np.median,'max', "count"]).apply(lambda x: round(x)).style.bar(subset = ["min", "mean", "median", "max"], align = "mid", color = rudá)

### extrém 12.000 !!! - vyškrtnuto

#data[data["cena_metr"] == 568.3]
#data[data["cena_metr"] == 1]


### 2.b) Ceny podle měst

In [ ]:
data["cena"].agg(['min', "mean", np.median,'max', "count"]).apply(lambda x: round(x))

In [ ]:
data[data["plocha"] != -1]["cena_metr"].agg(['min', "mean", np.median,'max', "count"]).apply(lambda x: round(x))


### 2.b.1) Nejdražší města

In [ ]:
data.groupby(["město", "okres"])["cena"].agg(["min",  np.median, "mean","max", "count"]).nlargest(15, "median").apply(lambda x: round(x)).style.bar(subset = ["mean", "median"], align = "mid", color = červená)

data[data["plocha"] != -1].groupby(["město", "okres"])["cena_metr"].agg(["min",  np.median, "mean","max", "count"]).nlargest(15, "median").apply(lambda x: round(x)).style.bar(subset = ["mean", "median"], align = "mid", color = červená)

# Subset jen říká, kam chci barvu

# Upravený kod, jde řadit nLargest i podle jednoho z více prvků !!
# upravený kod, jde formátovat i některé sloupce
# 
# V1 :nechám agg( [sloupec ]) a pak musím definovat který chci v Nlargest ale mám frame a nemusím dávat "to_frame()"
# V2: nechám jen agg(sloupec) a pak nemusím definovat podle kterého chci řadit ale zas musím udělat to_frame()

### 2.b.2) Nejlevnější města

In [ ]:
data.groupby(["město", "okres"])["cena"].agg(["min",  np.median, "mean","max", "count"]).sort_values( "median",ascending = False).tail(15).apply(lambda x: round(x)).style.bar(subset = ["mean", "median"], align = "mid", color = zelená)  # Subset jen říká, kam chci barvu

data[data["plocha"] != -1].groupby(["město", "okres"])["cena_metr"].agg(["min",  np.median, "mean","max", "count"]).sort_values( "median",ascending = False).tail(15).apply(lambda x: round(x)).style.bar(subset = ["mean", "median"], align = "mid", color = zelená)  # Subset jen říká, kam chci barvu

#  sort_values( ascending = False, "median").head(15)  mi nahrazuje    .nsmalest(15, "median"),  protože jsem chtěl aby to bylo seřazené v sestupném pořadí


### 2.c) Ceny podle okresů

### 2.c.1) Nejdražší okresy

In [ ]:
data.groupby("okres")["cena"].agg(['min', "mean", np.median,'max', "count"]).nlargest(15, "median").apply(lambda x: round(x)).style.bar(subset = ["mean", "median"], align = "mid", color = červená)

data[data["plocha"] != -1].groupby("okres")["cena_metr"].agg(['min', "mean", np.median,'max', "count"]).nlargest(15, "median").apply(lambda x: round(x)).style.bar(subset = ["mean", "median"], align = "mid", color = červená)


### 2.c.2) Nejlevnější okresy

In [ ]:
data.groupby("okres")["cena"].agg(['min', "mean", np.median,'max', "count"]).sort_values( "median",ascending = False).tail(15).apply(lambda x: round(x)).style.bar(subset = ["mean", "median"], align = "mid", color = zelená) 

data[data["plocha"] != -1].groupby("okres")["cena_metr"].agg(['min', "mean", np.median,'max', "count"]).sort_values( "median",ascending = False).tail(15).apply(lambda x: round(x)).style.bar(subset = ["mean", "median"], align = "mid", color = zelená) 


### 2.d) Ceny podle krajů

In [ ]:
data.groupby("kraj")["cena"].agg(['min', "mean", np.median,'max', "count"]).nlargest(15, "median").apply(lambda x: round(x)).style.bar(subset = ["mean", "median"], align = "mid", color = rudá)

data[data["plocha"] != -1].groupby("kraj")["cena_metr"].agg(['min', "mean", np.median,'max', "count"]).nlargest(15, "median").apply(lambda x: round(x)).style.bar(subset = ["mean", "median"], align = "mid", color = rudá)


### 3) Ceny podle místnosí PRAHA a MOST


### 3.a) Ceny podle místností PRAHA

In [ ]:
data[data["město"] == "Praha"].groupby("pokoje")["cena"].agg(['min', "mean", np.median,'max', "count"]).apply(lambda x: round(x)).style.bar(subset = ["min", "mean", "median", "max"], align = "mid", color = rudá)

data[(data["město"] == "Praha") & (data["plocha"] != -1)].groupby("pokoje")["cena_metr"].agg(['min', "mean", np.median,'max', "count"]).apply(lambda x: round(x)).style.bar(subset = ["min", "mean", "median", "max"], align = "mid", color = rudá)



# Postřeh: kde je větší nabídka, jsou i větší extrémy, aspoň co se  minim týče

## Postřeh: V praze byt za 320.000 a za 180.000 = Ve skutečnosti jen zprostředkování nájmu :/
# Vyškrtnuto

### 3.b) Ceny městských částí Prahy
* celkem 95 městských částí

### 3.b.1) Nejdražší pražské části

In [ ]:
data[data["město"] == "Praha"].groupby("oblast")["cena"].agg(['min', "mean", np.median,'max', "count"]).nlargest(15, columns = "median").apply(lambda x: round(x)).style.bar(subset = ["mean", "median"], align = "mid", color = červená)


data[(data["město"] == "Praha") & (data["plocha"] != -1)].groupby("oblast")["cena_metr"].agg(['min', "mean", np.median,'max', "count"]).nlargest(15, columns = "median").apply(lambda x: round(x)).style.bar(subset = ["mean", "median"], align = "mid", color = červená)


### 3.b.2) Nejlevnější pražské části

In [ ]:
data[data["město"] == "Praha"].groupby("oblast")["cena"].agg(['min', "mean", np.median,'max', "count"]).sort_values( "median",ascending = False).tail(15).apply(lambda x: round(x)).style.bar(subset = ["mean", "median"], align = "mid", color = zelená)

data[(data["město"] == "Praha") & (data["plocha"] != -1)].groupby("oblast")["cena_metr"].agg(['min', "mean", np.median,'max', "count"]).sort_values( "median",ascending = False).tail(15).apply(lambda x: round(x)).style.bar(subset = ["mean", "median"], align = "mid", color = zelená)


### 3.c) Ceny podle místností MOST

In [ ]:
#data[data["město"] == "Most"].groupby(["město", "pokoje"])["cena"].agg(['min', "mean", np.median,'max', "count"]).apply(lambda x: round(x)).style.bar(subset = ["min", "mean", "median", "max"], align = "mid", color = rudá)

#Ten poslední jediný záznam to trošku rozbourává ...Takže ho zkusím spíš odebrat
data[(data["město"] == "Most") & (data["pokoje"] != "4+kk")].groupby(["město", "pokoje"])["cena"].agg(['min', "mean", np.median,'max', "count"]).apply(lambda x: round(x)).style.bar(subset = ["min", "mean", "median", "max"], align = "mid", color = rudá)

data[(data["plocha"] != -1) &(data["město"] == "Most") & (data["pokoje"] != "4+kk")].groupby(["město", "pokoje"])["cena_metr"].agg(['min', "mean", np.median,'max', "count"]).apply(lambda x: round(x)).style.bar(subset = ["min", "mean", "median", "max"], align = "mid", color = rudá)



### 3.d)  PRAHA vs MOST

In [ ]:
# data[(data.město == "Praha") | (data.město == "Most") ].groupby(["pokoje","město" ])["cena"].agg(['min', "mean", np.median,'max', "count"]).apply(lambda x: round(x)).style.bar(subset = ["min", "mean", "median", "max"], align = "mid", color = rudá)

# Porovnám jen ty pokoje, které existují i v mostě, vyhnu se tím super extrémům
# Odebral jsem i 4+kk protože za Most byl jediný záznam - viz předešlý graf

data[((data.město == "Praha") | (data.město == "Most")) & (data.pokoje != "4+kk")& (data.pokoje != "5+1") & (data.pokoje != "5+kk") & (data.pokoje != "6-a-vice") & (data.pokoje != "atypicky") & (data.pokoje != "5+1") ]\
.groupby(["pokoje","město" ])["cena"].agg(['min', "mean", np.median,'max', "count"]).apply(lambda x: round(x)).style.bar(subset = ["min", "mean", "median", "max"], align = "mid", color = rudá)


data[(data["plocha"] != -1) & ((data.město == "Praha") | (data.město == "Most")) & (data.pokoje != "4+kk")& (data.pokoje != "5+1") & (data.pokoje != "5+kk") & (data.pokoje != "6-a-vice") & (data.pokoje != "atypicky") & (data.pokoje != "5+1") ]\
.groupby(["pokoje","město" ])["cena_metr"].agg(['min', "mean", np.median,'max', "count"]).apply(lambda x: round(x)).style.bar(subset = ["min", "mean", "median", "max"], align = "mid", color = rudá)


#data[data["plocha"] != -1]

### Zajímavý extrém... Praha 1+kk za 10 mega

In [ ]:
data[(data.město == "Praha") & (data.cena == 10390)  & (data.pokoje == "1+kk")]

### 4) Pivoty

### 4.a) Pivot Pražské části x Pokoje
* 95 pražských částí

In [ ]:
table = pd.pivot_table(data[data["město"] == "Praha"], values='cena', index=['oblast'],
                    columns=['pokoje'], aggfunc=np.mean).apply(lambda x: round(x))

# V1- Jen bílé nuly   # table.highlight_null(null_color='white')
# V2 - cmap = cm -> bílo-zelená    cm = sns.light_palette("green", as_cmap=True)
# V3 - cmap = "viridis"
# V4 - cmap = "magma"
# V5 - Inferno
# V6 - hot

# V5 - "magma_r"
table.style.background_gradient(cmap='magma_r')

### Nedodělaný HeatMap

In [ ]:
import matplotlib.pyplot as plt

fig, ax = plt.subplots(figsize=(15,35)) 

sns.heatmap(table, annot=True,fmt=".0f")

### 4.b) Pivot Kraje x Pokoje

In [ ]:
table = pd.pivot_table(data, values='cena', index=['kraj'],
                    columns=['pokoje'], aggfunc=np.mean).apply(lambda x: round(x))

cm = sns.light_palette("green", as_cmap=True)
table.style.background_gradient(cmap='magma_r')



### 4.c) Pivot Kraje (bez Prahy) x Pokoje

In [ ]:
table = pd.pivot_table(data[data.město != "Praha"], values='cena', index=['kraj'],
                    columns=['pokoje'], aggfunc=np.mean).apply(lambda x: round(x))

cm = sns.light_palette("green", as_cmap=True)
table.style.background_gradient(cmap='magma_r')

In [ ]:
def magnify():
    return [dict(selector="th",
                 props=[("font-size", "8pt")]),
            dict(selector="td",
                 props=[('padding', "0em 0em")]),
            dict(selector="th:hover",
                 props=[("font-size", "24pt")]),
            dict(selector="tr:hover td:hover",
                 props=[('max-width', '600px'),
                        ('font-size', '36pt')])
]

In [ ]:
table.style.background_gradient(cmap = "magma_r")\
    .set_properties(**{'max-width': '1200px', 'font-size': '1pt'})\
    .set_caption("Hover to magnify")\
    .set_precision(10)\
    .set_table_styles(magnify())

# Precision ovlivní délku čísla

## 5) Interaktívní widgety = DODĚLAT

In [ ]:
def unique_sorted_values_plus_ALL(array):
    unique = array.unique().tolist()
    unique.sort()
    return unique

cities = widgets.Dropdown(options =  unique_sorted_values_plus_ALL(data.city_one_praha))

def dropdown_year_eventhandler(change):
    display(data[data.city_one_praha == change.new])

cities.observe(dropdown_year_eventhandler, names='value')

display(cities)

 

In [ ]:
output = widgets.Output()
display(output)

In [ ]:
from ipywidgets import interact

import jupyterlab-widgets as lab_widgets

In [ ]:
#    jupyter labextension install @jupyter-widgets/jupyterlab-manager
# Nevidím nikde svj návod, ale jel jsme jinak podle: https://towardsdatascience.com/bring-your-jupyter-notebook-to-life-with-interactive-widgets-bc12e03f0916



ALL = 'ALL'
def unique_sorted_values_plus_ALL(array):
    unique = array.unique().tolist()
    unique.sort()
    unique.insert(0, ALL)
    return unique
################################

dropdown_year = widgets.Dropdown(options = unique_sorted_values_plus_ALL(data.city_one_praha))

output_year = widgets.Output()

def dropdown_year_eventhandler(change):
    output_year.clear_output()
    with output_year:
        if (change.new == ALL):
            display(data)
        else:
            display(data[data.city_one_praha == change.new])

dropdown_year.observe(dropdown_year_eventhandler, names='value')

display(dropdown_year)

        





In [ ]:
display(output_year)
